### PR TITLE
LwM2M suspend and  resume api

### PR DESCRIPTION
**Pull request by jheiskan81**
New API for suspend and resume LwM2M engine.

New event LWM2M_RD_CLIENT_EVENT_ENGINE_SUSPENDED for indicate application
that engine is suspended.

Simplify stack suspend and resume state same time for queue mode.

New CONFIG_LWM2M_RD_CLIENT_SUSPEND_SOCKET_AT_IDLE for enable skip socket close
at RX_OFF_IDLE state that socket is only suspended and close is called only
when connection is resumed.

### CODE REVIEW

File: subsys/net/lib/lwm2m/lwm2m_engine.c [hunk 1]
```diff
@@ -6178,6 +6227,21 @@ int lwm2m_engine_start(struct lwm2m_ctx *client_ctx)
...
+
+	k_thread_suspend(socket_loop_tid);
+}
+
+
+void lwm2m_engine_resume(void)
+{
+	k_thread_resume(socket_loop_tid);
+	LOG_INF("LWM2M engine socket receive thread resume ");
+
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_engine.c:6242]: Unneeded empty line.

File: subsys/net/lib/lwm2m/lwm2m_engine.c [hunk 2]
```diff
@@ -6189,7 +6253,7 @@ static int lwm2m_engine_init(const struct device *dev)
 	(void)memset(block1_contexts, 0, sizeof(block1_contexts));
 
 	/* start sock receive thread */
-	k_thread_create(&engine_thread_data, &engine_thread_stack[0],
+	socket_loop_tid = k_thread_create(&engine_thread_data, &engine_thread_stack[0],
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_engine.c:6256]: Perhaps `engine_thread_id` would be a more descriptive name?

File: subsys/net/lib/lwm2m/lwm2m_engine.c [hunk 3]
```diff
@@ -5833,6 +5890,9 @@ static void socket_loop(void)
 		}
 
 		for (i = 0; i < sock_nfds; i++) {
+
+			if (sock_ctx[i]->sock_fd < 0)
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_engine.c:5894]: Was this change intended? It seems a bit off.

File: subsys/net/lib/lwm2m/lwm2m_engine.c [hunk 4]
```diff
@@ -1450,28 +1455,104 @@ int lwm2m_init_message(struct lwm2m_message *msg)
...
+#endif
+		lwm2m_socket_update(client_ctx);
+		client_ctx->sock_fd = socket_temp_id;
 	}
+	return ret;
+}
+
+int lwm2m_engine_connection_resume(struct lwm2m_ctx *client_ctx)
+{
+
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_engine.c:1529]: Unneeded empty line.

File: subsys/net/lib/lwm2m/lwm2m_engine.c [hunk 5]
```diff
@@ -1450,28 +1455,104 @@ int lwm2m_init_message(struct lwm2m_message *msg)
...
+int lwm2m_close_suspend(struct lwm2m_ctx *client_ctx)
 {
-#ifdef CONFIG_LWM2M_DTLS_SUPPORT
-	if (!client_ctx->use_dtls) {
-		return 0;
+	int ret = 0;
+
+	if (client_ctx->sock_fd >= 0 && !client_ctx->connection_suspended) {
+		int socket_temp_id = client_ctx->sock_fd;
+
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_engine.c:1513]: Redundant empty line.

File: subsys/net/lib/lwm2m/lwm2m_rd_client.c [hunk 6]
```diff
@@ -55,9 +55,10 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 #include <errno.h>
 #include <zephyr/init.h>
 #include <zephyr/sys/printk.h>
-
+#include <net/socket.h>
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_rd_client.c:58]:
```suggestion
#include <zephyr/net/socket.h>
```

Is it really needed though here?

File: subsys/net/lib/lwm2m/lwm2m_engine.c [hunk 7]
```diff
@@ -1450,28 +1455,104 @@ int lwm2m_init_message(struct lwm2m_message *msg)
...
-int lwm2m_engine_connection_resume(struct lwm2m_ctx *client_ctx)
+		/* Enable Queue mode buffer store */
+		client_ctx->buffer_client_messages = true;
+#endif
+		lwm2m_socket_update(client_ctx);
+	}
+	return ret;
+}
+
+int lwm2m_close_suspend(struct lwm2m_ctx *client_ctx)
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_engine.c:1507]: Why `close_suspend`? Nothing seems to be closed in this case. Perhaps `lwm2m_socket_suspend()` would be enough?

File: subsys/net/lib/lwm2m/lwm2m_rd_client.c [hunk 8]
```diff
@@ -1091,6 +1097,44 @@ static void lwm2m_rd_client_service(struct k_work *work)
...
+				client.ctx->event_cb(client.ctx, event);
+			}
+			lwm2m_close_suspend(client.ctx);
+			suspended_client_state = get_sm_state();
+			client.engine_state = ENGINE_SUSPENDED;
+			client.suspend_thread = false;
+			k_mutex_unlock(&client.mutex);
+			while(get_sm_state() == ENGINE_SUSPENDED) {
+				k_sleep(K_MSEC(STATE_MACHINE_UPDATE_INTERVAL_MS / 2));
+			}
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_rd_client.c:1115]: I'm a bit confused again, isn't the engine thread supposed to be suspended at this point? Why is this loop here?

File: subsys/net/lib/lwm2m/Kconfig [hunk 9]
```diff
@@ -197,6 +197,12 @@ config LWM2M_TLS_SESSION_CACHING
 	help
 	  Enabling this only when feature is supported in TLS library.
 
+config LWM2M_RD_CLIENT_SUSPEND_SOCKET_AT_IDLE
+	bool "Socket close is skipped at RX_ON_IDLE state"
+	help
+	  This config is not close socket when LwM2M engine enter suspend'
+	  or RX_OFF_IDLE state.
```
[the remark this answers is not in the document]
Comment (jheiskan81) [reply subsys/net/lib/lwm2m/Kconfig:204]: OK I will fix this.

File: subsys/net/lib/lwm2m/lwm2m_engine.c [hunk 10]
```diff
@@ -1450,28 +1455,104 @@ int lwm2m_init_message(struct lwm2m_message *msg)
...
+		int socket_temp_id = client_ctx->sock_fd;
+
+
+		client_ctx->sock_fd = -1;
+		client_ctx->connection_suspended = true;
+#if defined(CONFIG_LWM2M_QUEUE_MODE_ENABLED)
+		/* Enable Queue mode buffer store */
+		client_ctx->buffer_client_messages = true;
+#endif
+		lwm2m_socket_update(client_ctx);
```
[the remark this answers is not in the document]
Comment (jheiskan81) [reply subsys/net/lib/lwm2m/lwm2m_engine.c:1521]: OK- let me re-think can I avoid this operation.

File: subsys/net/lib/lwm2m/lwm2m_engine.c (hunk 3 above)
Comment (jheiskan81) [reply subsys/net/lib/lwm2m/lwm2m_engine.c:5894]: yes true. I need to check. I think there should be continue inside.

File: subsys/net/lib/lwm2m/lwm2m_engine.c [hunk 11]
```diff
@@ -6178,6 +6227,21 @@ int lwm2m_engine_start(struct lwm2m_ctx *client_ctx)
 	return lwm2m_socket_start(client_ctx);
 }
 
+void lwm2m_engine_pause(void)
+{
+	LOG_INF("LWM2M engine socket receive thread paused");
+
+	k_thread_suspend(socket_loop_tid);
```
[the remark this answers is not in the document]
Comment (jheiskan81) [reply subsys/net/lib/lwm2m/lwm2m_engine.c:6234]: currently RD Client is calling this. Is calling this at own thread. API call must be call from different thread I think.

File: subsys/net/lib/lwm2m/lwm2m_rd_client.c (hunk 6 above)
Comment (jheiskan81) [reply subsys/net/lib/lwm2m/lwm2m_rd_client.c:58]: OK I can remove and test.

File: subsys/net/lib/lwm2m/lwm2m_rd_client.c (hunk 8 above)
Comment (jheiskan81) [reply subsys/net/lib/lwm2m/lwm2m_rd_client.c:1115]: I change a design so don't worry.

**Comment by jheiskan81** [comment]
@rlubos I have been updated your raised issues and changes pause and resume functionality. @SeppoTakalo Please review.

File: subsys/net/lib/lwm2m/Kconfig [hunk 12]
```diff
@@ -197,6 +197,12 @@ config LWM2M_TLS_SESSION_CACHING
 	help
 	  Enabling this only when feature is supported in TLS library.
 
+config LWM2M_RD_CLIENT_SUSPEND_SOCKET_AT_IDLE
```
[the remark this answers is not in the document]
Comment (SeppoTakalo) [reply subsys/net/lib/lwm2m/Kconfig:197]: I wonder if we should drop the whole `LWM2M_RD_CLIENT` Kconfig. I don't believe the engine could work anymore without it. I also don't think anyone test or use it in that configuration. I really cannot see the point.

However, not in this PR though..

File: subsys/net/lib/lwm2m/lwm2m_rd_client.c [hunk 13]
```diff
@@ -201,6 +198,15 @@ static void set_sm_state(uint8_t sm_state)
...
 		client.ctx->event_cb(client.ctx, event);
 	}
+
+	/* Suspend socket after Event callback */
+	if (event == LWM2M_RD_CLIENT_EVENT_QUEUE_MODE_RX_OFF) {
+#if defined(CONFIG_LWM2M_RD_CLIENT_SUSPEND_SOCKET_AT_IDLE)
+		lwm2m_socket_suspend(client.ctx);
+#else
+		lwm2m_close_socket(client.ctx);
+#endif
```
Comment (SeppoTakalo) [inline subsys/net/lib/lwm2m/lwm2m_rd_client.c:208]:
```suggestion
		if (IS_ENABLED(CONFIG_LWM2M_RD_CLIENT_SUSPEND_SOCKET_AT_IDLE)) {
				lwm2m_socket_suspend(client.ctx);
		} else { 
				lwm2m_close_socket(client.ctx);
		}

```

**Review by SeppoTakalo — DISMISSED** [review]
Looking good.
Just some styling suggestions.

File: subsys/net/lib/lwm2m/Kconfig (hunk 12 above)
[the remark this answers is not in the document]
Comment (rlubos) [reply subsys/net/lib/lwm2m/Kconfig:197]: Not in this PR, but I agree that it'd make sense to drop this config.

File: subsys/net/lib/lwm2m/lwm2m_engine.c [hunk 14]
```diff
@@ -161,26 +166,103 @@ char *sprint_token(const uint8_t *token, uint8_t tkl)
...
-#endif
+
 	return 0;
 }
-#endif
-#if defined(CONFIG_LWM2M_QUEUE_MODE_ENABLED)
+
 int lwm2m_push_queued_buffers(struct lwm2m_ctx *client_ctx)
 {
+	#if defined(CONFIG_LWM2M_QUEUE_MODE_ENABLED)
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_engine.c:265]: Preprocessor code shouldn't be indented.

File: subsys/net/lib/lwm2m/lwm2m_engine.c [hunk 15]
```diff
@@ -192,9 +274,9 @@ int lwm2m_push_queued_buffers(struct lwm2m_ctx *client_ctx)
 		msg = SYS_SLIST_CONTAINER(msg_node, msg, node);
 		sys_slist_append(&msg->ctx->pending_sends, &msg->node);
 	}
+	#endif
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_engine.c:277]: Ditto

File: subsys/net/lib/lwm2m/lwm2m_engine.c [hunk 16]
```diff
@@ -487,6 +532,18 @@ int lwm2m_socket_add(struct lwm2m_ctx *ctx)
...
+static void lwm2m_socket_update(struct lwm2m_ctx *ctx)
+{
+	for (int i = 0; i < sock_nfds; i++) {
+		if (sock_ctx[i] != ctx) {
+			continue;
+		}
+		sock_fds[i].fd = ctx->sock_fd;
+		return;
+	}
+
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_engine.c:544]: Redundant empty line.

File: subsys/net/lib/lwm2m/lwm2m_shell.c [hunk 17]
```diff
@@ -433,6 +435,22 @@ static int cmd_update(const struct shell *sh, size_t argc, char **argv)
 	return 0;
 }
 
+static int cmd_pause(const struct shell *sh, size_t argc, char **argv)
+{
+	ARG_UNUSED(argc);
+	ARG_UNUSED(argv);
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_shell.c:441]: `sh` also seems to be unused.

File: subsys/net/lib/lwm2m/lwm2m_shell.c [hunk 18]
```diff
@@ -433,6 +435,22 @@ static int cmd_update(const struct shell *sh, size_t argc, char **argv)
...
+
+	return lwm2m_engine_pause();
+}
+
+static int cmd_resume(const struct shell *sh, size_t argc, char **argv)
+{
+	ARG_UNUSED(argc);
+	ARG_UNUSED(argv);
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_shell.c:449]: `sh` also seems to be unused.

File: subsys/net/lib/lwm2m/lwm2m_engine.c [hunk 19]
```diff
@@ -859,6 +923,44 @@ int lwm2m_engine_start(struct lwm2m_ctx *client_ctx)
...
 }
 
+int lwm2m_engine_pause(void)
+{
+	char buffer[32];
+	const char *str;
+
+	str = k_thread_state_str(engine_thread_id, buffer, sizeof(buffer));
+	if (suspend_engine_thread || !strcmp(str, "suspended")) {
+		LOG_WRN("Engine Already at Suspended");
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_engine.c:933]: Just a suggestion:
```suggestion
		LOG_WRN("Engine thread already suspended");
```

**Review by rlubos — COMMENTED** [review]
Just a few nits.

**Comment by jheiskan81** [comment]
@SeppoTakalo & @rlubos Thanks for feedback. I have been fixed raised items now.

**Review by rlubos — APPROVED** [review]

**Review by SeppoTakalo — APPROVED** [review]
